### PR TITLE
Make agent local state async

### PR DESCRIPTION
**Pull request by ShimmerGlass**
Local filesystem operations can fail during service registration /
deregistration and leave the agent in an invalid state.
This change makes this operation async and have them retried
periodically if they fail.
In addition to the consistency benefits, this moves the code from
agent.go to the local package and mock it during tests, allowing to
properly clean the TestAgent datadirs on shutdown.

### CODE REVIEW

File: agent/local/peristent_state.go [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+			if !ok {
+				time.AfterFunc(retryInterval, s.triggerSync)
+			}
+		}
+	}()
+
+	return s
+}
+
+func (s *PersistentState) WriteService(def PersistedService) {
```
Comment (pierresouchay) [inline agent/local/peristent_state.go]: API Doc?

**Comment by pierresouchay** [comment]
@Aestek Can we ensure this one can be merged against current master?

**Comment by i0rek** [comment]
@ShimmerGlass This seems interesting from the description and I would like to review it if you are still interested in adding it! Could you rebase it before so that I can see what is going on?

**Comment by dnephin** [comment]
Thank you for working on this improvement! Some additional structure around local agent persistent would be super valuable.

Unfortunately large PRs like this age very poorly, and at this point it seems like rebasing is probably more work than attempting to re-apply the changes. Since this PR is not merge-able and is not being worked on I'm going to close it.

If anyone is interested in making a change like this I'd be happy to work with you to try and find ways to make the change in a series of smaller PRs. Smaller diffs should make it easier for us to review quickly, and will reduce the number of conflicts when a PR does take a little while to get merged.

**Pull request closed without merging**